libraries/Microsoft.Bot.Builder.Dialogs/WaterfallDialog.cs

```diff
@@ -69,6 +69,10 @@ public override async Task<DialogTurnResult> DialogResumeAsync(DialogContext dc,
 
             // Increment step index and run step
             var state = dc.ActiveDialog.State;
+
+            // For issue https://github.com/Microsoft/botbuilder-dotnet/issues/871
+            // See the linked issue for details the stepIndex is stored as an Int32 in CosmosDB
+            // which was being cast to int (Int64). This change ensures the correct cast was done.
```

Comment:
This comment is probably not correct. 

See the C# Language spec here:
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/types#simple-types

You can see that "int' and "int32" are aliases for each other. 

With that said, "Convert" and "Cast" often have different exception related behavior, so I think a bit more digging is still needed. 

Can you dig in a bit more and see?

Reply:
Yeah technically it's not a Cosmos DB problem; it's a JSON deserialization issue. This could coming out of a text file on disk and it would be the same problem. I explained it in more detail in the linked GH issue.

Reply:
Sure thing.  My understanding was that when you cast using (int)object is that the bits are inspected and the best type is determined from the bits. I think this is also what @drub0y was getting at more elegantly in his explanation.  In this case, it was Int64 which I was trying to relay in my comments by saying `int (Int64)` but I guess it didn't come across.  I was trying to be brief to spare everyone from reading a novel in my comments.  I can be more explicit.  I understand that the issue does in fact not lie with CosmosDB, but I also wanted the reader to understand that this is an issue with that particular data store, again a symptom of my brevity.  I'll edit and clarify.  Thanks for the feedback.